### DISCUSSION TITLE
Extract data from proof images using AI

### DISCUSSION THREAD

**Discussion by raphodn** [openfoodfacts/open-prices#560] [General]
There is lots of interesting info in the proof images : 
- price tag : price, barcode
- receipt : multiple prices, no barcodes nor EAN but a "product name", date, location
- exif data : date & location ?

What if the server was able to do some image recognition, and return to the frontend all of that info ? :star_struck: 

#### Using GenAI

- [x] extract data from price tags (using Google's Gemini) (#557)

@TTalex has built a prototype, using Google's Gemini API.

#### Using an external library

- [ ] #670

#### Our own trained model

- [x] Proof classification : integrated into Robotoff (not yet into OP) (#394)
- [x] Google Cloud vision : we call the API on each new image : useful for moderation (e.g. detect inappropriate images) (#320)
- [x] Extract price tags (bounding boxes) from Proof images : #611
- [x] extract data from price tags

@raphael0202 is working on building a long-term solution based on training data.

#### Receipt

Without the product EAN or barcode, how to link to the exact product in the OFF database ? 

### Related discussions

- #523
- https://github.com/openfoodfacts/open-prices/discussions/526#discussioncomment-11323772

### Related issues

- #320
- #394
- #308
- #511
- #569
- #611

**Comment by teolemon** [comment]
On this one, it's probably about building the endpoints 
- EXIF to (timestamp + lists of candidate stores)
- Image to (list of candidates barcodes + candidate prices)

Then those endpoints can be reused on proof addition, or for retro-processing

**Comment by raphodn** [comment]
On Slack you mentionned a tool/workflow you used to extract barcode info for instance.

We could use this issue to list available tools ?

I've been thinking of ways to accelerate data collection, taking pictures of shelves, and having an image recognition tool that would return a list of product codes+prices.

Then would come a step to integrate the "chosen" tool in the backend ?

**Comment by teolemon** [comment]
- There are Tensorflow.JS based solutions
 - https://blog.tensorflow.org/2022/06/ocr-in-browser-using-tensorflowjs.html
 - https://demo-doctr-tensorflowjs.mindee.com/
 - Tesseract.js https://tesseract.projectnaptha.com/
- We can offload processing to Robotoff, which has all the tooling in place
- We can do it in the Open Prices backend, it's not rocket-science (possibly with the Robotoff lib)

**Comment by teolemon** [comment]
For EXIF to stores
- https://github.com/ozgecinko/image-metadata-extractor
- https://codereview.stackexchange.com/questions/221809/get-nearest-major-city-of-an-image-using-exif-gps-data
- https://medium.com/geekculture/extract-gps-information-from-photos-using-python-79288c58ccd9
- Client side: https://github.com/exif-js/exif-js

**Comment by alexfauquette** [comment]
I gave a try few days ago with terseract, but surprisingly it has lots of false positive, and is not perfect on barcode (some digits are missing)

I need to try again with different resolution

![image](https://github.com/openfoodfacts/open-prices/assets/45398769/f28b0f56-6e2e-459f-98ab-e59c98456e7b)

**Comment by teolemon** [comment]
Or we could cloudvision our way out of it, it's not like we'll be out of credits this year.

**Comment by teolemon** [comment]
- We could just use Cloud Vision #320

**Comment by raphodn** [comment]
Image examples

|Price tag|Multiple price tags|Receipt|
|---|---|---|
|![05JD3g310N](https://github.com/openfoodfacts/open-prices/assets/7147385/dad5a307-2348-48aa-88b4-05d002e089ce)|![IMG_20240127_172140](https://github.com/openfoodfacts/open-prices/assets/7147385/0a73de7b-6b74-40aa-955b-b0b286ba986d)|![receipt](https://github.com/openfoodfacts/open-prices/assets/7147385/dffcee51-9fa9-4e19-87ec-64ff759e7a06)|
